p2p: Delay querying DNS seeds

ajtowns opened the pull request:
Changes the logic for querying DNS seeds: after this PR, if there's less than 1000 entries in addrman, it will still usually query DNS seeds after 11s (unless the first few peers tried mostly succeed), but if there's more than 1000 entries it won't try DNS seeds until 5 minutes have passed without getting multiple outbound peers. (If there's 0 entries in addrman, it will still immediately query the DNS seeds). Additionally, delays querying DNS seeds while the p2p network is not active.

Fixes #15434

ajtowns commented:
Somewhat follows the suggestions of [@luke-jr](https://github.com/bitcoin/bitcoin/pull/15558#issuecomment-471239506) and [@jgarzik](https://github.com/bitcoin/bitcoin/pull/15558#issuecomment-471208937) in #15558

ajtowns commented:
Behaviour looks like:

```
2019-09-23T10:54:49Z dnsseed thread start
2019-09-23T10:54:49Z Waiting 300 seconds before falling back to DNS seeds.
2019-09-23T10:54:50Z trying connection XXX lastseen=662.4hrs
2019-09-23T10:54:55Z connection to XXX timeout
2019-09-23T10:54:55Z trying connection XXX lastseen=197.4hrs
2019-09-23T10:54:56Z Added connection peer=0
2019-09-23T10:54:56Z New outbound peer connected: ...
2019-09-23T10:54:56Z trying connection XXX lastseen=469.4hrs
```

DNS thread would previously have decided to contact seeds here.
```
2019-09-23T10:55:01Z connection to XXX timeout
2019-09-23T10:55:02Z trying connection XXX lastseen=685.1hrs
2019-09-23T10:55:07Z connection to XXX timeout
2019-09-23T10:55:07Z trying connection XXX lastseen=232.5hrs
2019-09-23T10:55:12Z connection to XXX timeout
2019-09-23T10:55:13Z trying connection XXX lastseen=388.0hrs
```

DNS thread checks for early exit here, but doesn't need one so keeps waiting.

```
2019-09-23T10:55:18Z connection to XXX timeout
2019-09-23T10:55:18Z trying connection XXX lastseen=15.6hrs
2019-09-23T10:55:19Z Added connection peer=1
2019-09-23T10:55:19Z New outbound peer connected: ...
2019-09-23T10:55:19Z trying connection XXX lastseen=36.8hrs
2019-09-23T10:55:24Z connection to XXX timeout
2019-09-23T10:55:25Z trying connection XXX lastseen=214.8hrs
2019-09-23T10:55:30Z connection to XXX timeout
2019-09-23T10:55:30Z trying connection XXX lastseen=524.4hrs
2019-09-23T10:55:35Z connection to XXX timeout
2019-09-23T10:55:36Z trying connection XXX lastseen=1705.3hrs
```

DNS thread checks again for early exit, and since there's two connections now, does:

```
2019-09-23T10:55:39Z P2P peers available. Skipped DNS seeding.
2019-09-23T10:55:39Z dnsseed thread exit
```

practicalswift commented:
Concept ACK: seed querying should be kept to a minimum for obvious reasons

naumenkogs commented:
Concept ACK.

This doesn't make us depend on a single seed, does it?
Also, perhaps move some of the magic numbers to constants?

ajtowns commented:
Fixed the logic to make it query in batches of 3 more sensibly, and bumped the delay for adding fixed seeds to 6 minutes from startup (fixed nodes are only added if addrman is empty, which will only happen if DNS seeds aren't replying -- but this may be due to the first one we try hanging, rather than all of them failing, so 6m is a compromise).

Behaviour for empty peers.dat, but no ability to actually connect anywhere is to immediately query 3 random dns seeds, then wait 11 seconds and try the next 3, etc:

```
2019-09-23T18:02:46Z Loaded 0 addresses from peers.dat  0ms
2019-09-23T18:02:46Z dnsseed thread start
2019-09-23T18:02:46Z Loading addresses from DNS seed dnsseed.bluematt.me
2019-09-23T18:02:47Z Loading addresses from DNS seed dnsseed.bitcoin.dashjr.org
2019-09-23T18:02:52Z Loading addresses from DNS seed seed.bitcoin.jonasschnelli.ch
2019-09-23T18:02:52Z Waiting 11 seconds before querying DNS seeds.
2019-09-23T18:03:03Z Loading addresses from DNS seed dnsseed.emzy.de
2019-09-23T18:03:03Z Loading addresses from DNS seed seed.bitcoin.sipa.be
2019-09-23T18:03:03Z Loading addresses from DNS seed seed.bitcoin.sprovoost.nl
2019-09-23T18:03:03Z Waiting 11 seconds before querying DNS seeds.
2019-09-23T18:03:14Z Loading addresses from DNS seed seed.btc.petertodd.org
2019-09-23T18:03:14Z 33 addresses found from DNS seeds
2019-09-23T18:03:14Z dnsseed thread exit
```

Same deal if there's some but less than 1k entries in addrman, except it'll wait 11s first:

```
2019-09-23T18:23:15Z Loaded 33 addresses from peers.dat  0ms
2019-09-23T18:23:15Z dnsseed thread start
2019-09-23T18:23:15Z Waiting 11 seconds before querying DNS seeds.
2019-09-23T18:23:26Z Loading addresses from DNS seed dnsseed.bitcoin.dashjr.org
2019-09-23T18:23:26Z Loading addresses from DNS seed dnsseed.emzy.de
2019-09-23T18:23:26Z Loading addresses from DNS seed dnsseed.bluematt.me
2019-09-23T18:23:27Z Waiting 11 seconds before querying DNS seeds.
2019-09-23T18:23:38Z Loading addresses from DNS seed seed.btc.petertodd.org
2019-09-23T18:23:38Z Loading addresses from DNS seed seed.bitcoin.sipa.be
2019-09-23T18:23:38Z Loading addresses from DNS seed seed.bitcoin.jonasschnelli.ch
2019-09-23T18:23:38Z Waiting 11 seconds before querying DNS seeds.
2019-09-23T18:23:49Z Loading addresses from DNS seed seed.bitcoin.sprovoost.nl
2019-09-23T18:23:49Z 0 addresses found from DNS seeds
2019-09-23T18:23:49Z dnsseed thread exit
```

(seed.bitcoinstats.com seems to hang for me, so I've dropped it from my chainparams, but that change isn't in this patch)

ariard commented on src/net.cpp:
```diff
@@ -53,6 +53,15 @@ static constexpr int DUMP_PEERS_INTERVAL = 15 * 60;
+/** How long to delay before querying DNS seeds
+ */
+static constexpr std::chrono::seconds DNSSEEDS_DELAY_FEW_PEERS{11};
+static constexpr std::chrono::seconds DNSSEEDS_DELAY_MANY_PEERS{300};
```
nit: // 5 minutes, also move them to net.h ?

ajtowns replied:
They're not needed from other modules, so doesn't make sense to have them in net.h

ariard commented on src/net.cpp:
```diff
@@ -1543,30 +1552,55 @@ void CConnman::ThreadDNSAddressSeed()
+    //   reduces traffic to the seeds.
+    // * When querying DNS seeds query a few at once, this ensures
+    //   that we don't give DNS seeds the ability to eclipse nodes
+    //   that query them.
```
You can lay out trade-off further: "Every new DNS seed queried increases the bar for eclipsing node but leaks more privacy."

ariard commented on src/net.cpp:
```diff
@@ -1543,30 +1552,55 @@ void CConnman::ThreadDNSAddressSeed()
     Shuffle(seeds.begin(), seeds.end(), rng);
     int seeds_right_now = 0; // Number of seeds left before testing if we have enough connections
     int found = 0;
+    const std::chrono::seconds seeds_wait_time = (addrman.size() >= 1000 ? DNSSEEDS_DELAY_MANY_PEERS : DNSSEEDS_DELAY_FEW_PEERS);
```
What the logic for picking 1000 ? If we assume uniform address propagation, what's the ratio of reliable public nodes on others in addrman and what the probability of not connecting successfully to at least 3 outbound peers ? Based on that, I would say intuitively we may scale down number a bit to increase number of nodes who delay DNS query.

ajtowns replied:
1000 is the maximum number of entries the `addr` message can include, so seems a good threshold to indicate you've managed to pick up some live data from the p2p network about potential peers, rather than just some results from DNS lookups.

ariard replied:
Okay, make sense given we send `GETADDR` at `VERSION` reception which means if you have at only one reliable outbound connection you should be able to bootstrap your addrman without seed assistance.

ariard commented on src/net.cpp:
```diff
@@ -1543,30 +1552,55 @@ void CConnman::ThreadDNSAddressSeed()
-                LogPrintf("P2P peers available. Skipped DNS seeding.\n");
-                return;
+            LogPrintf("Waiting %d seconds before querying DNS seeds.\n", seeds_wait_time.count());
+            std::chrono::seconds to_wait = seeds_wait_time;
```
For fresh nodes, addrman can't be bigger than 1000 after querying 3 dns (because nMaxIPs == 256), but if we change DNSSEEDS_TO_QUERY_AT_ONCE, we may assign  seeds_wait_time before every time evaluation instead

ajtowns replied:
With the current code:

 * if addrman starts off at 0, it'll query all DNS seeds with no delay
 * if addrman starts off under 1000, it'll query all DNS seeds in groups of 3 after waiting 11 seconds before each group
 * if addrman starts off over 1000, it'll query all DNS seeds in groups of 3 after waiting 5 minutes before each group

That seems mostly fine to me?

(The timeout for connecting to a peer is nConnectTimeout which is 5 seconds; so 11 seconds will pass if you happen to try three peers that timeout before getting two that succeed)

ariard replied:
Right, I think I forget we set `seeds_right_now` to number of seeds if if we don't have any known peers (IIRC my comment motivation)

ariard commented on src/net.cpp:
```diff
@@ -1543,30 +1552,55 @@ void CConnman::ThreadDNSAddressSeed()
+                }
             }
             seeds_right_now += DNSSEEDS_TO_QUERY_AT_ONCE;
+        } else if (seeds_right_now == 0) {
```
Not sure about this, I would say that's a behavior change. Previously node with  an empty addrman would have query all DNS seeds due to seeds_right_now == -1 as you noticed in #15558. Now you may exit early, if among DNSSEEDS_TO_QUERY_AT_ONCE you find at least 2 outbound peers but with final result of having less address origin diversity in your addrman..

ariard commented on src/net.cpp:
```diff
@@ -1715,7 +1749,7 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
 
         // Add seed nodes if DNS seeds are all down (an infrastructure attack?).
-        if (addrman.size() == 0 && (GetTime() - nStart > 60)) {
+        if (addrman.size() == 0 && (GetTime() - nStart > FIXEDSEEDS_DELAY)) {
```
This is a behavior change too, I do understand the logic to increase privacy for fresh nodes if DNS  are busy but how did you pick up this number ?

ajtowns replied:
This change isn't needed -- if addrman is going to end up empty, it will have been empty for each dns seed which means there would have been no delay in querying the seed.

ariard reviewed (commented):
@ajtowns what's the state of this PR ?

I think that's interesting work but title/description should be updated to inform about the 3 behavior changes:

- delay querying DNS seeds for 5min if addrman is less than 1000 (concept ACK)
- reduce set of queried DNS for fresh nodes to 3 instead of all (need to think more)
- increase fallback to hardcoded seeds for fresh nodes to 6 min instead of 60s (concept ACK)

gmaxwell commented:
This shouldn't run if the node is not currently making automatic connections (network disable or -noconnect).

ajtowns commented:
@gmaxwell

> This shouldn't run if the node is not currently making automatic connections (network disable or -noconnect).

For `-noconnect` we already have "parameter interaction: -connect set -> setting -dnsseed=0" which prevents the dns thread from ever starting. Not sure what network disable means, should we not do dns seeding if `-onlynet=onion` or something?

ajtowns committed:
DNS seeds: wait for 5m instead of 11s if 1000+ peers are known

If 1000 potential peers are known, wait for 5m before querying DNS seeds
for more peers, since eventually the addresses we already know should
get us connected. Also check every 11s whether we've got enough active
outbounds that DNS seeds aren't worth querying, and exit the dnsseed
thread early if so.

ajtowns commented:
Rebased, reverted delay for fixed seeds, and changed to explicitly query all dns seeds if we start up with an empty addrman.

MarcoFalke commented:
> Not sure what network disable means

There is a `setnetworkactive` RPC, which can disable the connman

ajtowns committed:
DNS seeds: don't query DNS while network is inactive

ajtowns commented:
Added patch to avoid querying DNS seeds while `setnetworkactive` is set to false.

hebasto commented:
Concept ACK.

hebasto commented on src/net.cpp:
```diff
@@ -1573,26 +1579,55 @@ void CConnman::ThreadDNSAddressSeed()
+            LogPrintf("Waiting %d seconds before querying DNS seeds.\n", seeds_wait_time.count());
+            std::chrono::seconds to_wait = seeds_wait_time;
+            while (to_wait.count() > 0) {
+                std::chrono::seconds w{25}; // check for early abort every 25 seconds
```
As 25 seconds is a timeout for the _early_ exit from dnsseed thread, we could do not introduce a new heuristic constant at all:
```diff
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -41,9 +41,9 @@
 static_assert(MINIUPNPC_API_VERSION >= 10, "miniUPnPc API version >= 10 assumed");
 #endif
 
-#include <unordered_map>
-
+#include <chrono>
 #include <math.h>
+#include <unordered_map>
 
 // Dump addresses to peers.dat every 15 minutes (900s)
 static constexpr int DUMP_PEERS_INTERVAL = 15 * 60;
@@ -1601,12 +1601,12 @@ void CConnman::ThreadDNSAddressSeed()
     for (const std::string& seed : seeds) {
         if (addrman.size() > 0 && seeds_right_now == 0) {
             LogPrintf("Waiting %d seconds before querying DNS seeds.\n", seeds_wait_time.count());
-            std::chrono::seconds to_wait = seeds_wait_time;
+            std::chrono::milliseconds to_wait{seeds_wait_time};
             while (to_wait.count() > 0) {
-                std::chrono::seconds w{25}; // check for early abort every 25 seconds
-                if (w > to_wait) w = to_wait;
-                if (!interruptNet.sleep_for(w)) return;
-                to_wait -= w;
+                std::chrono::milliseconds recheck_period{nConnectTimeout};
+                if (recheck_period > to_wait) recheck_period = to_wait;
+                if (!interruptNet.sleep_for(recheck_period)) return;
+                to_wait -= recheck_period;
 
                 int nRelevant = 0;
                 {
```

ajtowns replied:
It'd make sense for `DNSSEEDS_DELAY_FEW_PEERS` to be calculated from that rather than a constant, but the early exit ought to be based on how long it takes for a connection to _succeed_ rather than the timeout time

ajtowns replied:
On the other hand, can easily reuse `DNSSEEDS_DELAY_FEW_PEERS` instead of a new constant, so will just do that.

jgarzik commented:
concept ACK

When DNS seeding was introduced, the idea was to be a backup, for when P2P peer addr exchange is not working for whatever reason.  It seems preferable to make the polling closer to "24 hours" or "never" in the case where AddrMan is healthy and happy.

Regardless of that preference, concept ACK as stated above, as this appears to be a strict improvement & resolve an issue.

hebasto commented on src/net.cpp:
```diff
@@ -1573,26 +1579,55 @@ void CConnman::ThreadDNSAddressSeed()
             }
             seeds_right_now += DNSSEEDS_TO_QUERY_AT_ONCE;
+        } else if (seeds_right_now == 0) {
+            seeds_right_now += DNSSEEDS_TO_QUERY_AT_ONCE;
```
It seems this branch will never be executed since all of the dns seeds are explicitly queried if we start up with an empty addrman:
https://github.com/bitcoin/bitcoin/blob/5bf16653168cbc3c8bd0814cfa937d0dc8a879ff/src/net.cpp#L1582-L1585

ajtowns replied:
Rearranged to `if (s_r_n == 0) { s_r_n += AT_ONCE; if (addrman.size() == 0) { wait(); } }`

hebasto commented on src/net.cpp:
```diff
@@ -1573,26 +1579,55 @@ void CConnman::ThreadDNSAddressSeed()
+                        nRelevant += pnode->fSuccessfullyConnected && !pnode->fFeeler && !pnode->fOneShot && !pnode->m_manual_connection && !pnode->fInbound;
+                    }
+                }
+                if (nRelevant >= 2) {
```
While here mind replacing a magic number 2 with a named constant?

ariard commented:
@ajtowns I think you need to update commit message, it's still mentioning the 25s delays while you now rely on DNSSEED_DELAY_FEW_PEERS?

sipa commented on src/net.cpp:
```diff
@@ -1573,31 +1579,68 @@ void CConnman::ThreadDNSAddressSeed()
+                LogPrintf("Waiting %d seconds before querying DNS seeds.\n", seeds_wait_time.count());
+                std::chrono::seconds to_wait = seeds_wait_time;
+                while (to_wait.count() > 0) {
+                    std::chrono::seconds w = DNSSEEDS_DELAY_FEW_PEERS;
```
```suggestion
                    std::chrono::seconds w = std::min(DNSSEEDS_DELAY_FEW_PEERS, to_wait);
```

ajtowns replied:
Took this since I rebased to update the commit description

sipa reviewed (commented):
utACK b1338cc25c0d66a73aea760117e488b6c91d0d43

hebasto commented on src/net.cpp:
```diff
@@ -1573,31 +1579,67 @@ void CConnman::ThreadDNSAddressSeed()
+    // * If we continue having problems, eventually query all the
+    //   DNS seeds, and if that fails too, also try the fixed seeds.
+    //   (done in ThreadOpenConnections)
+    const std::chrono::seconds seeds_wait_time = (addrman.size() >= DNSSEEDS_DELAY_PEER_THRESHOLD ? DNSSEEDS_DELAY_MANY_PEERS : DNSSEEDS_DELAY_FEW_PEERS);
```
Could add `#include <chrono>`? See [Source code organization](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#source-code-organization).

ajtowns replied:
We already use std::chrono in net.cpp; I think it's included via net.h via addrman.h via util/system.h via util/time.h. So yeah, could see the value in cleaning that up, but don't think it needs to be in this patch.

hebasto commented on src/net.cpp:
```diff
@@ -1573,31 +1579,67 @@ void CConnman::ThreadDNSAddressSeed()
-            if (nRelevant >= 2) {
-                LogPrintf("P2P peers available. Skipped DNS seeding.\n");
-                return;
+            if (addrman.size() > 0) {
```
Considering https://github.com/bitcoin/bitcoin/blob/96954d17948662672cababc940e453dff08e8cbb/src/net.cpp#L1582-L1585 and https://github.com/bitcoin/bitcoin/blob/96954d17948662672cababc940e453dff08e8cbb/src/net.cpp#L1602

this point is not reachable when `addrman.size() == 0`, and `addrman.size() > 0` will always be evaluated to true, so it could be dropped, no?

ajtowns replied:
If we did reach it with `addrman.size() == 0`, waiting would be pointless, so I don't think the logic is wrong; there's no `else` so there's no unreachable code, just a conditional branch that always happens to go one way. I think double checking is a bit more robust in case we change the code to querying some subset of dns seeds when addrman is empty in future, eg if we had dozens of dns seeds available.

Sjors replied:
> in case we change the code to querying some subset of dns seeds when addrman is empty in future

That's worth explaining in a comment. It may also be more readable to have a `bool skip_wait`

hebasto commented on src/net.cpp:
```diff
@@ -1573,31 +1579,67 @@ void CConnman::ThreadDNSAddressSeed()
+                            nRelevant += pnode->fSuccessfullyConnected && !pnode->fFeeler && !pnode->fOneShot && !pnode->m_manual_connection && !pnode->fInbound;
+                        }
+                    }
+                    if (nRelevant >= 2) {
```
```suggestion
                    if (nRelevant >= DNSSEEDS_ENOUGH_PEERS) {
```

ajtowns replied:
I don't really think this adds anything -- the logs in the body of the if seems like it documents the value better than giving it a name would, and we don't reuse it anywhere else.

hebasto reviewed (changes requested):
Tested 96954d17948662672cababc940e453dff08e8cbb on Linux Mint 19.3: works as expected.

ariard reviewed (approved):
Tested ACK 96954d1, on Debian 9.1. Both MANY_PEERS/FEW_PEERS cases work.

```
2020-04-22T22:13:40Z Loaded 3579 addresses from peers.dat  10ms
2020-04-22T22:13:40Z init message: Starting network threads...
2020-04-22T22:13:40Z net thread start
2020-04-22T22:13:40Z addcon thread start
2020-04-22T22:13:40Z Using /16 prefix for IP bucketing.
2020-04-22T22:13:40Z init message: Done loading
2020-04-22T22:13:40Z msghand thread start
2020-04-22T22:13:40Z dnsseed thread start
2020-04-22T22:13:40Z Waiting 300 seconds before querying DNS seeds.
2020-04-22T22:18:40Z Loading addresses from DNS seed dnsseed.bitcoin.dashjr.org
2020-04-22T22:18:40Z Loading addresses from DNS seed seed.bitcoinstats.com
2020-04-22T22:18:41Z Loading addresses from DNS seed seed.bitcoin.sipa.be
```

naumenkogs reviewed (approved):
utACK 96954d17948662672cababc940e453dff08e8cbb
Feel free to ignore comment suggestions if you’re not force-pushing any other fixes.

ajtowns commented on src/net.cpp:
```diff
@@ -1573,31 +1579,67 @@ void CConnman::ThreadDNSAddressSeed()
     if (gArgs.GetBoolArg("-forcednsseed", DEFAULT_FORCEDNSSEED)) {
         // When -forcednsseed is provided, query all.
         seeds_right_now = seeds.size();
+    } else if (addrman.size() == 0) {
```
It tells us we don't have any known peers (there's already a comment to that effect) -- perhaps this is the first time you've started bitcoind or you've just restarted after deleting peers.dat. If you previously queried the dns seeds, or have at some point connected to some other peers that aren't in IBD, this shouldn't be true any more.

MarcoFalke commented:
@ajtowns Would you prefer to add a new commit with documentation or instead have this merged first or something else?

ajtowns commented:
> @ajtowns Would you prefer to add a new commit with documentation or instead have this merged first or something else?

Maybe a separate PR for updating the comments? See #19084

Sjors commented on src/net.cpp:
```diff
@@ -1573,31 +1579,61 @@ void CConnman::ThreadDNSAddressSeed()
+                LogPrintf("Waiting %d seconds before querying DNS seeds.\n", seeds_wait_time.count());
+                std::chrono::seconds to_wait = seeds_wait_time;
+                while (to_wait.count() > 0) {
+                    std::chrono::seconds w = std::min(DNSSEEDS_DELAY_FEW_PEERS, to_wait);
```
Could use a comment of why you're sleeping in shorter intervals: because we can exit this thread once we have enough nodes.

Sjors reviewed (approved):
tACK 96954d1 (rebased on master) on macOS 10.15.4. It found it useful to run with `-debug=addrman` and change `DNSSEEDS_DELAY_MANY_PEERS` to something lower to test the behaviour, as well as renaming `peers.dat` to test the peer threshold.

The second commit can be tested by calling `bitcoin-cli setnetworkactive false` within `DNSSEEDS_DELAY_MANY_PEERS` seconds of the DNS thread starting.

Consider adding the improved documentation as a new commit on top of this PR. That doesn't break any reviews, but it makes it easier to squash if you need to change anything else.

ajtowns commented on src/net.cpp:
```diff
@@ -1573,31 +1579,61 @@ void CConnman::ThreadDNSAddressSeed()
         seeds_right_now = seeds.size();
+    } else if (addrman.size() == 0) {
+        // If we have no known peers, query all.
+        seeds_right_now = seeds.size();
```
In #15558 the first run will set `seeds_right_now = 0` then see that `addrman.size() > 0 && seeds_right_now == 0` is false (since `addrman.size() == 0`) so will not sleep nor bump `seeds_right_now` by `DNSSEEDS_TO_QUERY_AT_ONCE`, it will then query a seed, and decrement `seeds_right_now` so that it's `-1`, at which point `addrman.size() > 0 && seeds_right_now == 0` will continue being false (because `seeds_right_now < 0`) and will query all seeds without waiting between them.

This retains that behaviour, but is explicit about it. See [suhas's comment on 15558](https://github.com/bitcoin/bitcoin/pull/15558#discussion_r327643907)

fanquake reviewed (approved):
ACK 96954d17948662672cababc940e453dff08e8cbb - Ran some tests of different scenarios. More documentation is being added in #19084.

```bash
# already had a large peers.dat
src/bitcoind | rg dnsseed
2020-05-29T10:00:29.745627Z [dnsseed] dnsseed thread start
2020-05-29T10:00:29.745657Z [dnsseed] Waiting 300 seconds before querying DNS seeds.
2020-05-29T10:00:40.745989Z [dnsseed] P2P peers available. Skipped DNS seeding.
2020-05-29T10:00:40.746071Z [dnsseed] dnsseed thread exit

# deleted peers.dat
src/bitcoind | rg dnsseed
2020-05-29T10:01:38.673439Z [dnsseed] dnsseed thread start
2020-05-29T10:01:38.673508Z [dnsseed] Loading addresses from DNS seed dnsseed.bitcoin.dashjr.org
2020-05-29T10:01:38.674948Z [dnsseed] Loading addresses from DNS seed dnsseed.emzy.de
2020-05-29T10:01:39.681687Z [dnsseed] Loading addresses from DNS seed dnsseed.bluematt.me
2020-05-29T10:01:40.081659Z [dnsseed] Loading addresses from DNS seed seed.bitcoinstats.com
2020-05-29T10:01:40.705362Z [dnsseed] Loading addresses from DNS seed seed.bitcoin.jonasschnelli.ch
2020-05-29T10:01:41.014162Z [dnsseed] Loading addresses from DNS seed seed.bitcoin.sprovoost.nl
2020-05-29T10:01:42.539538Z [dnsseed] Loading addresses from DNS seed seed.btc.petertodd.org
2020-05-29T10:01:45.568217Z [dnsseed] Loading addresses from DNS seed seed.bitcoin.sipa.be
2020-05-29T10:01:45.570067Z [dnsseed] 305 addresses found from DNS seeds
2020-05-29T10:01:45.570103Z [dnsseed] dnsseed thread exit

# peers.dat < 1000 addresses
src/bitcoind | rg dnsseed
2020-05-29T10:04:03.566819Z [dnsseed] dnsseed thread start
2020-05-29T10:04:03.566983Z [dnsseed] Waiting 11 seconds before querying DNS seeds.
2020-05-29T10:04:14.567539Z [dnsseed] P2P peers available. Skipped DNS seeding.
2020-05-29T10:04:14.567643Z [dnsseed] dnsseed thread exit

# -connect=0
src/bitcoind -connect=0 | rg dnsseed
2020-05-29T10:05:19.064542Z [init] InitParameterInteraction: parameter interaction: -connect set -> setting -dnsseed=0

# called setnetworkactive false early
src/bitcoind | rg dnsseed
2020-05-29T10:07:31.399566Z [dnsseed] dnsseed thread start
2020-05-29T10:07:31.399595Z [dnsseed] Waiting 300 seconds before querying DNS seeds.
2020-05-29T10:12:31.466326Z [dnsseed] Waiting for network to be reactivated before querying DNS seeds.
2020-05-29T10:14:20.753947Z [dnsseed] Loading addresses from DNS seed dnsseed.emzy.de
2020-05-29T10:14:20.762381Z [dnsseed] Loading addresses from DNS seed dnsseed.bitcoin.dashjr.org
2020-05-29T10:14:20.763729Z [dnsseed] Loading addresses from DNS seed seed.bitcoin.jonasschnelli.ch
2020-05-29T10:14:20.765102Z [dnsseed] Waiting 300 seconds before querying DNS seeds.
2020-05-29T10:14:42.766527Z [dnsseed] 113 addresses found from DNS seeds
2020-05-29T10:14:42.766629Z [dnsseed] P2P peers available. Finished DNS seeding.
2020-05-29T10:14:42.766673Z [dnsseed] dnsseed thread exit

src/bitcoind | rg dnsseed
2020-05-29T10:15:02.325759Z [dnsseed] dnsseed thread start
2020-05-29T10:15:02.325804Z [dnsseed] Waiting 300 seconds before querying DNS seeds.
2020-05-29T10:15:13.326011Z [dnsseed] P2P peers available. Skipped DNS seeding.
2020-05-29T10:15:13.326123Z [dnsseed] dnsseed thread exit
```
